ambry-store/src/main/java/com/github/ambry/store/RemoteTokenTracker.java

```diff
@@ -15,29 +15,62 @@
 
 import com.github.ambry.clustermap.ReplicaId;
 import com.github.ambry.replication.FindToken;
+import com.github.ambry.utils.CrcInputStream;
+import com.github.ambry.utils.CrcOutputStream;
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.Utils;
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * {@link RemoteTokenTracker} tracks tokens from all peer replicas and updates them when handling metadata request from
  * peer node.
+ * We persist the peer replicas' token and the timestamp when the token is updated to a file.
  */
-public class RemoteTokenTracker {
+public class RemoteTokenTracker implements Closeable {
```

Comment:
Curious: What was the issue with the previous remote token tracker?

Reply:
Nothing is wrong with the previous tracker.

We have delete record tombstone issue. As long as there is one replica down, we cannot move forward to remove any tombstone because we don't know the replication token of that replica.
After we persist the token, if there is a server down, at least we can still continue to remove the delete record as long as the delete record is before the token.

Later, we can have more improvement. With the timestamp, we know if the replica is down for 1 month, we'll ignore that down replica. But this is not in the PR.